src/lobpcg/algorithm.rs

```diff
@@ -0,0 +1,658 @@
+/// as it could be usable. If there is no result at all, then `LobpcgResult::NoResult` is returned.
+/// This happens if the algorithm fails in an early stage, for example if the matrix `A` is not SPD
+#[derive(Debug)]
+pub enum LobpcgResult<A> {
+    Ok(Array1<A>, Array2<A>, Vec<A>),
+    Err(Array1<A>, Array2<A>, Vec<A>, LinalgError),
+    NoResult(LinalgError),
+}
+
+/// Solve the generalized eigenvalue problem with pencil (A, B)
+fn generalized_eig<S: Data<Elem = A>, A: NdFloat>(
+    a: ArrayBase<S, Ix2>,
+    b: ArrayBase<S, Ix2>,
+) -> Result<(Array1<A>, Array2<A>)> {
+    let (vals_b, vecs_b) = b.eigh()?;
+    let vals_b_recip = vals_b.mapv(|x| (x.max(A::from(1e-10f32).unwrap())).sqrt().recip());
+    let vecs_b_tilde = &vecs_b * &vals_b_recip;
+    let a_tilde = vecs_b_tilde.t().dot(&a.dot(&vecs_b_tilde));
+    let (vals_a, vecs_a) = a_tilde.eigh()?;
+    let vecs = vecs_b_tilde.dot(&vecs_a);
+
+    Ok((vals_a, vecs))
+}
+
+/// Solve full eigenvalue problem, sort by `order` and truncate to `size`
+fn sorted_eig<S: Data<Elem = A>, A: NdFloat>(
+    a: ArrayBase<S, Ix2>,
+    b: Option<ArrayBase<S, Ix2>>,
+    size: usize,
+    order: &Order,
+) -> Result<(Array1<A>, Array2<A>)> {
+    let n = a.len_of(Axis(0));
+
+    let res = match b {
+        Some(b) => generalized_eig(a, b)?,
+        _ => a.eigh()?.sort_eig(false),
+    };
+
+    // sort and ensure that signs are deterministic
+    let (vals, vecs) = res.sort_eig(false);
+    let s = vecs.row(0).mapv(|x| x.signum());
+    let vecs = vecs * &s;
+
+    Ok(match order {
+        Order::Largest => (
+            vals.slice_move(s![n-size..; -1]),
+            vecs.slice_move(s![.., n-size..; -1]),
+        ),
+        Order::Smallest => (vals.slice_move(s![..size]), vecs.slice_move(s![.., ..size])),
+    })
+}
+
+/// Masks a matrix with the given `matrix`
+fn ndarray_mask<A: NdFloat>(matrix: ArrayView2<A>, mask: &[bool]) -> Array2<A> {
+    assert_eq!(mask.len(), matrix.ncols());
+
+    let indices = (0..mask.len())
+        .zip(mask.iter())
+        .filter(|(_, b)| **b)
+        .map(|(a, _)| a)
+        .collect::<Vec<usize>>();
+
+    matrix.select(Axis(1), &indices)
+}
+
+/// Applies constraints ensuring that a matrix is orthogonal to it
+///
+/// This functions takes a matrix `v` and constraint-matrix `y` and orthogonalize `v` to `y`.
+fn apply_constraints<A: NdFloat>(
+    mut v: ArrayViewMut<A, Ix2>,
+    cholesky_yy: &Array2<A>,
+    y: ArrayView2<A>,
+) {
+    let gram_yv = y.t().dot(&v);
+
+    /*
+    let u = gram_yv
+        .columns()
+        .into_iter()
+        .map(|x| {
+            let res = cholesky_yy.solve_triangular(&x, UPLO::Upper)?;
+
+            res.to_vec()
+        })
+        .flatten()
+        .collect::<Vec<A>>();
+
+    let rows = gram_yv.len_of(Axis(0));
+    let u = Array2::from_shape_vec((rows, u.len() / rows), u).unwrap();*/
+    let u = cholesky_yy.solve_triangular(&gram_yv, UPLO::Lower).unwrap();
+
+    v -= &(y.dot(&u));
```

Comment:
`ndarray::linalg::general_mat_mul(-1, y, u, 1, &mut v)` for a way to do this operation without allocating

Reply:
okay I do this, but it requires a good documentation then

Reply:
Yeah just comment it with `v -= y . u` or something